cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRequestQueue.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.preview;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.app.preview.PreviewRequest;
+import io.cdap.cdap.app.preview.PreviewRequestQueue;
+import io.cdap.cdap.app.preview.PreviewRequestQueueState;
+import io.cdap.cdap.app.store.preview.PreviewStore;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.ApplicationId;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Thread-safe implementation of {@link PreviewRequestQueue} backed by {@link PreviewStore}.
+ */
+@ThreadSafe
+public class DefaultPreviewRequestQueue implements PreviewRequestQueue {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPreviewRequestQueue.class);
+  private final PreviewStore previewStore;
+  private final ConcurrentLinkedDeque<PreviewRequest> requestQueue;
+  private final int capacity;
+  private final long waitTimeOut;
+  private final AtomicInteger queueSize;
+
+  @Inject
+  public DefaultPreviewRequestQueue(CConfiguration cConf, PreviewStore previewStore) {
+    this.previewStore = previewStore;
+    this.capacity = cConf.getInt(Constants.Preview.WAITING_QUEUE_CAPACITY, 50);
+    this.waitTimeOut = cConf.getLong(Constants.Preview.WAITING_QUEUE_TIMEOUT_SECONDS, 60);
+    this.requestQueue = new ConcurrentLinkedDeque<>();
+    this.queueSize = new AtomicInteger();
```

Comment:
Why we need to track it separated from the `requestQueue.size()`? It is better to use the queue size as the only source of truth.

Reply:
fixed. Also made add method as `synchronized`. Did it in branch but moved that change to this PR.